pkg/cmd/cli/cmd/importimage.go

```diff
@@ -249,151 +249,134 @@ func (o *ImportImageOptions) waitForImport(resourceVersion string) (*imageapi.Im
-			From: kapi.ObjectReference{
-				Kind: "DockerImage",
-				Name: from,
-			},
-			ImportPolicy: imageapi.TagImportPolicy{Insecure: insecure},
+		// importing the entire repository
+		isi, err = o.importAll(stream)
+		if err != nil {
+			return nil, nil, err
 		}
 	} else {
-		isi.Spec.Images = append(isi.Spec.Images, imageapi.ImageImportSpec{
-			From: kapi.ObjectReference{
-				Kind: "DockerImage",
-				Name: from,
-			},
-			To:           &kapi.LocalObjectReference{Name: tag},
-			ImportPolicy: imageapi.TagImportPolicy{Insecure: insecure},
-		})
+		// importing a single tag
+		isi, err = o.importTag(stream)
+		if err != nil {
+			return nil, nil, err
+		}
 	}
 
 	return stream, isi, nil
 }
 
-func (o *ImportImageOptions) newImageStream() (string, *imageapi.ImageStream) {
+func (o *ImportImageOptions) newImageStream() (*imageapi.ImageStream, *imageapi.ImageStreamImport) {
 	from := o.From
 	tag := o.Tag
 	if len(from) == 0 {
 		from = o.Target
 	}
 	var stream *imageapi.ImageStream
+	// create new ImageStream
 	if o.All {
 		stream = &imageapi.ImageStream{
 			ObjectMeta: kapi.ObjectMeta{Name: o.Name},
 			Spec:       imageapi.ImageStreamSpec{DockerImageRepository: from},
 		}
-	}
-	stream = &imageapi.ImageStream{
-		ObjectMeta: kapi.ObjectMeta{Name: o.Name},
-		Spec: imageapi.ImageStreamSpec{
-			Tags: map[string]imageapi.TagReference{
-				tag: {
-					From: &kapi.ObjectReference{
-						Kind: "DockerImage",
-						Name: from,
+	} else {
+		stream = &imageapi.ImageStream{
+			ObjectMeta: kapi.ObjectMeta{Name: o.Name},
+			Spec: imageapi.ImageStreamSpec{
+				Tags: map[string]imageapi.TagReference{
+					tag: {
+						From: &kapi.ObjectReference{
+							Kind: "DockerImage",
+							Name: from,
+						},
 					},
 				},
 			},
-		},
+		}
+	}
+	// and accompanying ImageStreamImport
+	var isi *imageapi.ImageStreamImport
+	if o.All {
+		isi = o.newImageStreamImportAll(stream, from)
+	} else {
+		isi = o.newImageStreamImportTags(stream, map[string]string{tag: from})
 	}
-	return from, stream
+
+	return stream, isi
 }
 
-func (o *ImportImageOptions) importAll(stream *imageapi.ImageStream) (string, error) {
+func (o *ImportImageOptions) importAll(stream *imageapi.ImageStream) (*imageapi.ImageStreamImport, error) {
 	from := o.From
+	// update ImageStream appropriately
 	if len(from) == 0 {
-		if len(stream.Spec.DockerImageRepository) == 0 {
-			// FIXME soltysh: support all spec.Tags
-			return "", fmt.Errorf("flag --all is applicable only to images with spec.dockerImageRepository defined")
+		if len(stream.Spec.DockerImageRepository) != 0 {
+			from = stream.Spec.DockerImageRepository
+		} else {
+			tags := make(map[string]string)
+			for name, tag := range stream.Spec.Tags {
+				if tag.From != nil && tag.From.Kind == "DockerImage" {
+					tags[name] = tag.From.Name
+				}
+			}
+			if len(tags) == 0 {
+				return nil, fmt.Errorf("image stream does not have importable tags (pointing to DockerImage)")
```

Comment:
Apparently it'll be fixed/changed, due to https://bugzilla.redhat.com/show_bug.cgi?id=1343681. But that will be rather a followup to this one. For now I'll go with: 'image stream does not have tags pointing to external docker images'